sdks/python/apache_beam/runners/portability/artifact_service.py

```diff
@@ -0,0 +1,122 @@
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import hashlib
+import os
+import random
+import re
+
+from apache_beam.io import filesystems
+from apache_beam.portability.api import beam_artifact_api_pb2
+from apache_beam.portability.api import beam_artifact_api_pb2_grpc
+
+
+class BeamFilesystemArtifactService(
+    beam_artifact_api_pb2_grpc.ArtifactStagingServiceServicer,
+    beam_artifact_api_pb2_grpc.ArtifactRetrievalServiceServicer):
+
+  _DEFAULT_CHUNK_SIZE = 10 << 20  # 10mb
+
+  def __init__(self, root, chunk_size=_DEFAULT_CHUNK_SIZE):
+    self._root = root
+    self._chunk_size = chunk_size
+
+  def _sha256(self, string):
+    return hashlib.sha256(string.encode('utf-8')).hexdigest()
+
+  def _mkdir(self, retrieval_token):
+    dir = os.path.join(self._root, retrieval_token)
+    if not filesystems.FileSystems.exists(dir):
+      try:
+        filesystems.FileSystems.mkdirs(dir)
+      except Exception:
+        pass
+
+  def retrieval_token(self, staging_session_token):
+    return self._sha256(staging_session_token)
+
+  def _artifact_path(self, retrieval_token, name):
+    # These are user-provided, best to check.
+    assert re.match('^[0-9a-f]+$', retrieval_token)
+    self._mkdir(retrieval_token)
+    return os.path.join(self._root, retrieval_token, self._sha256(name))
+
+  def _manifest_path(self, retrieval_token):
+    # These are user-provided, best to check.
+    assert re.match('^[0-9a-f]+$', retrieval_token)
+    self._mkdir(retrieval_token)
+    return os.path.join(self._root, retrieval_token, 'manifest.proto')
+
+  def PutArtifact(self, request_iterator, context=None):
+    first = True
+    for request in request_iterator:
+      if first:
+        first = False
+        metadata = request.metadata.metadata
+        retrieval_token = self.retrieval_token(
+            request.metadata.staging_session_token)
+        self._mkdir(retrieval_token)
+        temp_path = os.path.join(
+            self._root,
+            retrieval_token,
+            '%x.tmp' % random.getrandbits(128))
+        fout = filesystems.FileSystems.create(temp_path)
+        hasher = hashlib.sha256()
+      else:
+        hasher.update(request.data.data)
+        fout.write(request.data.data)
+    fout.close()
+    data_hash = hasher.hexdigest()
+    if metadata.sha256 and metadata.sha256 != data_hash:
+      filesystems.FileSystems.delete([temp_path])
+      raise ValueError('Bad metadata hash: %s vs %s' % (
+          metadata.metadata.sha256, data_hash))
+    filesystems.FileSystems.rename(
+        [temp_path], [self._artifact_path(retrieval_token, metadata.name)])
+    return beam_artifact_api_pb2.PutArtifactResponse()
+
+  def CommitManifest(self, request, context=None):
+    retrieval_token = self.retrieval_token(request.staging_session_token)
+    with filesystems.FileSystems.create(
+        self._manifest_path(retrieval_token), 'w') as fout:
```

Comment:
Do we want 'wb'?

Reply:
Actually, we don't want anything here, given we're using Filesystems.create.